### PR TITLE
feat: add `ap` method to `Effect`, `ap` and `zipWith` to `Either` ⚡️

### LINKED ISSUE #1502: From Discord: Lack of `ap` Method in Either and Effect
# Summary
The user suggests that the `ap` method is missing in the `Either` and `Effect` types and believes it may have been overlooked because it is not commonly used.

# Discord thread

https://discord.com/channels/795981131316985866/1160871851631849532

### PR DESCRIPTION
**Pull request by kutyel**
Fixes #1502

### CODE REVIEW

File: src/Effect.ts [hunk 1]
```diff
@@ -4744,6 +4744,18 @@ export const zipWith: {
...
 } = fiberRuntime.zipWithOptions
 
+// -------------------------------------------------------------------------------------
+// applicatives
+// -------------------------------------------------------------------------------------
+/**
+ * @category combining
+ * @since 2.0.0
+ */
+export const ap: {
```
Comment (gcanti) [inline src/Effect.ts:4754]: the signature should be

```ts
export const ap: {
  <R2, E2, A>(that: Effect<R2, E2, A>): <R, E, B>(self: Effect<R, E, (a: A) => B>) => Effect<R | R2, E | E2, B>
  <R, E, A, B, R2, E2>(self: Effect<R, E, (a: A) => B>, that: Effect<R2, E2, A>): Effect<R | R2, E | E2, B>
} = dual(
  2,
  <R, E, A, B, R2, E2>(self: Effect<R, E, (a: A) => B>, that: Effect<R2, E2, A>): Effect<R | R2, E | E2, B> =>
    zipWith(self, that, (f, a) => f(a))
)
```

**Review by gcanti — CHANGES_REQUESTED** [review]

File: src/Effect.ts (hunk 1 above)
Comment (kutyel) [reply src/Effect.ts:4754]: Applied in https://github.com/Effect-TS/effect/pull/1504/commits/123a835195ad9476cc954b8a53a3099e672f928b, thanks @gcanti 🚀

**Review by gcanti — CHANGES_REQUESTED** [review]
I would add a few tests for the 3 new APIs.

p.s.
Could you please also run `pnpm docgen` to generate the updated docs?

File: test/Effect/applicative.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import * as Effect from "effect/Effect"
+import * as Fiber from "effect/Fiber"
+import * as Queue from "effect/Queue"
+import { assert, describe } from "vitest"
+
+describe.concurrent("Queue", () => {
+  it.effect("applicative take and offer", () =>
+    Effect.gen(function*($) {
+      const queue = yield* $(Queue.bounded<string>(100))
+      const fiber = yield* $(Queue.take(queue), Effect.ap(Queue.take(queue), (a, b) => a + b), Effect.fork)
```
Comment (kutyel) [inline test/Effect/applicative.ts:11]: @gcanti the test works but TS blows in this line, I guess I need to type cast to something more specific? 🤔

File: test/Effect/applicative.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import * as it from "effect-test/utils/extend"
+import * as Effect from "effect/Effect"
+import * as Fiber from "effect/Fiber"
+import * as Queue from "effect/Queue"
+import { assert, describe } from "vitest"
+
+describe.concurrent("Queue", () => {
```
Comment (gcanti) [inline test/Effect/applicative.ts:7]: typo: should be "Effect"

File: test/Effect/applicative.ts (hunk 2 above)
Comment (gcanti) [reply test/Effect/applicative.ts:11]: Not sure, but honestly l would switch to a simpler test using `Effect.succeed` and `Effect.fail`, along the lines of what we did with `Either`

**Review by gcanti — APPROVED** [review]

File: test/Effect/applicative.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    }))
+
+  it.effect("one failure in data-last position should fail", () =>
+    Effect.gen(function*($) {
+      const result = yield* $(
+        Effect.succeed(add).pipe(Effect.ap(Effect.succeed(1)), Effect.ap(Effect.fail("c"))),
+        Effect.match({
+          onFailure: Either.left,
+          onSuccess: Either.right
+        })
```
Comment (gcanti) [inline test/Effect/applicative.ts:22]: FYI this can be replaced with `Effect.either`

**Pull request merged**